trestle/core/rules_validator.py

```diff
@@ -0,0 +1,119 @@
+# limitations under the License.
+"""Validate by confirming rule parameter values are consistent."""
+import logging
+import pathlib
+from typing import List
+
+from trestle.common.common_types import TopLevelOscalModel
+from trestle.common.common_types import TypeWithSetParams
+from trestle.common.list_utils import as_list, deep_set, get_item_from_list
+from trestle.core.catalog.catalog_interface import CatalogInterface
+from trestle.core.profile_resolver import ProfileResolver
+from trestle.core.validator import Validator
+from trestle.oscal.ssp import ByComponent, ImplementedRequirement, Statement, SystemComponent, SystemSecurityPlan
+
+logger = logging.getLogger(__name__)
+
+
+class RulesValidator(Validator):
+    """Validator to confirm all rule parameter values are consistent."""
+
+    rule_param_values_dict = {}
+
+    def iter_and_add(
+        self,
+        by_components: List[ByComponent],
+        components: List[SystemComponent],
+        cat_int: CatalogInterface,
+        imp_requirement: ImplementedRequirement,
+        statement: Statement = None
+    ) -> None:
+        """
+        Iterate all by components in an object and add the rule shared parameter values to list.
+
+        args:
+            by_components: A list of by components.
+            components: Components defined at System implementation level.
+            cat_int: Instance of catalog interface with controls catalog loaded.
+            imp_requirement: Current implemented requirement.
+            statement: Statement if iterating by statements.
+        """
+        for by_component in as_list(by_components):
+            component = get_item_from_list(components, by_component.component_uuid, lambda x: x.uuid)
+            # first adds implmented requirement set params
+            if not statement:
+                self.add_shared_param(imp_requirement, imp_requirement.control_id, cat_int, component.title)
+            # then  adds by components set params per requirement
+            self.add_shared_param(by_component, imp_requirement.control_id, cat_int, component.title)
+
+    def add_shared_param(
+        self, item: TypeWithSetParams, control_id: str, cat_int: CatalogInterface, comp_name: str = ''
+    ) -> None:
+        """
+        Add a rule shared parameter to the rule shared parameters list.
+
+        args:
+            item: Generic item to iterate over parameters.
+            control_id: Current control id.
+            cat_int: Instance of catalog interface with controls catalog loaded.
+            comp_name: Component name to save.
+        """
+        for set_param in as_list(item.set_parameters):
+            control = cat_int.get_control_by_param_id(set_param.param_id)
+            if not control:
+                deep_set(self.rule_param_values_dict, [set_param.param_id, comp_name, control_id], set_param.values)
+
+    def model_is_valid(self, model: TopLevelOscalModel, quiet: bool, trestle_root: pathlib.Path = '') -> bool:
```

Comment:
Sorry I have just realised there is one major issue with this PR.

`trestle validate` does not pass `trestle_root` to this validator. Please see line 71 in `validator.py`.

Here you initialise it to the empty string. This will fail when user runs `trestle validate` and no guidance would be provided to the user on how to fix this issue.

You can either take a current repository as the root or raise an exception to notify user that root is required. In any case even if the user provides the `--trestle_root` it will not be passed.

I would recommend creating a test that would run `trestle validate` on the ssp repository using `execute_command_and_assert` to catch this. 

Also please give it a test on `ocp-oscal-ssp`

Reply:
Thanks @enikonovad this is something special to validating the ssp that is a break from the normal validation calls.  @AleJo2995 sorry I didn't anticipate this issue, but it shouldn't be too hard to fix.  Normally a model in memory can be validated without trestle root because it doesn't need to load anything from the trestle workspace.  But the ssp validation is special since it needs to load the profile.  I think the simplest fix is to change the signature of `model_is_valid` in the base class so it passes in an Optional trestle root as `pathlib.Path` or None.  None of the other validators need trestle_root but the ssp one does.  Also not that pathlib.Path is a special class and is not a string - so if not present it should default to None.